sqrl-tools/sqrl-auth/src/main/java/com/datasqrl/auth/AuthProvider.java

```diff
@@ -0,0 +1,155 @@
+package com.datasqrl.auth;
+
+import static com.datasqrl.auth.AuthUtils.*;
+
+import com.datasqrl.util.FileUtil;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URLEncoder;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.MessageDigest;
+import java.security.SecureRandom;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.Base64;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+import lombok.SneakyThrows;
+
+public class AuthProvider {
+
+  private static final Path DATASQRL_CONFIG_DIRECTORY = FileUtil.getUserRoot().resolve(".datasqrl");
+  private static final Path AUTH_TOKEN_PATH = DATASQRL_CONFIG_DIRECTORY.resolve("auth");
+
+  private static final long TIMEOUT_IN_SECONDS = 60;
+
+  private final String state = getRandomString(16);
+  private final String codeVerifier = getRandomString(43);
+  private final String codeChallenge = generateCodeChallenge(codeVerifier);
+
+  private final HttpClient client;
+
+    public AuthProvider(HttpClient client) {
+        this.client = client;
+    }
+
+    @SneakyThrows
+  public String loginToRepository() {
+    CompletableFuture<String> tokenFuture = new CompletableFuture<>();
+
+    Vertx vertx = Vertx.vertx();
+    vertx.deployVerticle(new OAuthCallbackVerticle(code -> exchangeCodeForToken(code, tokenFuture)));
```

Comment:
should be in the try block so vertx can clean up if something goes wrong with deploying the verticle or opening the browser